commit 2d85c8ddc44e267c4c6d645a8c13aa50ff94d20f (refs/changes/83/3583/2)
Author: Austin Wise <AustinWise@gmail.com>
Date:   2018-03-12T13:23:22-07:00 (1 year, 7 months ago)
    
    joyent/illumos-joyent#163 fix stack selection for delivery of Linux signals

diff --git a/usr/src/lib/brand/lx/lx_brand/common/signal.c b/usr/src/lib/brand/lx/lx_brand/common/signal.c
index a8e3601cb9..5ae399bcd8 100644
--- a/usr/src/lib/brand/lx/lx_brand/common/signal.c
+++ b/usr/src/lib/brand/lx/lx_brand/common/signal.c
@@ -26,6 +26,7 @@
 
 /*
  * Copyright 2018 Joyent, Inc. All rights reserved.
+ * Copyright 2018 Austin Wise <austin@awise.us>
  */
 
 #include <sys/types.h>
@@ -1636,8 +1637,10 @@ lx_sigdeliver(int lx_sig, siginfo_t *sip, ucontext_t *ucp, size_t stacksz,
 	 * stack specified by the user.
 	 */
 	newstack = (lxsap->lxsa_flags & LX_SA_ONSTACK) &&
-	    !(lxtsd->lxtsd_sigaltstack.ss_flags & (LX_SS_ONSTACK |
-	    LX_SS_DISABLE));
+	    !(lxtsd->lxtsd_sigaltstack.ss_flags & LX_SS_DISABLE) &&
+	    (orig_sp < (uintptr_t)lxtsd->lxtsd_sigaltstack.ss_sp ||
+	    orig_sp >= (uintptr_t)(lxtsd->lxtsd_sigaltstack.ss_sp +
+	    lxtsd->lxtsd_sigaltstack.ss_size));
 
 	/*
 	 * Find the first unused region of the Linux process stack, where
